### ISSUE TITLE
MoE kernel

### ISSUE THREAD

**Issue by ByronHsu** [linkedin/Liger-Kernel#206]
### 🚀 The feature, motivation and pitch

Currently the most popular library might be https://github.com/databricks/megablocks. Would be interesting if we can implement it in triton and make it HF compatible

**Comment by S1ro1** [comment]
Will do a research on this more, if anyone has any insights on what could/should be implemented, resp. details on to how, cc me.

**Comment by yundai424** [comment]
@S1ro1 one straightforward idea is to parallelize expert forward (just like what megablock impl does). Right now in HF model code the MoE block is performed sequentiallyexpert-by-expert

**Comment by S1ro1** [comment]
@yundai424 Haven't seen one either, gonna try patching either Mixtral or Nllb with our kernels and profile it, will decide on what to do after that I guess.
Implementing dMoE (dropless MoE) could also be interesting. Will try to send the profiler benchmarks tomorrow and could discuss more in depth. Also I suppose Mixtral > Nllb.

**Comment by pramodith** [comment]
@yundai424 @S1ro1 I'd like to help with this, but wanted to pin down some of the exact steps that can be taken to make the MoE layer more efficient. 

Per my understanding the HF implementation of [Mixtral](https://github.com/huggingface/transformers/blob/fc1ae7f30f1d16c7652c28dd8d91c5d8a8ed2f15/src/transformers/models/mixtral/modeling_mixtral.py#L661), calls the experts sequentially because each expert can be allocated a variable number of tokens and they wanted to avoid dropping any tokens.

I guess we can start off by implementing `ParallelMLP`[code](https://github.com/databricks/megablocks/blob/84286de8ab5be0c73928a0059f50c7e2b650e4b1/megablocks/layers/moe.py#L96) in this repo, but I'm not sure if this'll actually include any new triton kernels. Most of the logic there seems to deal with sharding and distributing the required tensors across ranks.

**Comment by S1ro1** [comment]
@pramodith I totally agree with starting with the MLP, however i'm currently surprisingly swamped with school so I won't have time to collaborate on this. So feel free to take this.
